File: app/src/main/java/com/example/exchangingprivatelessons/domain/model/Lesson.kt
package com.example.exchangingprivatelessons.domain.model

import com.google.firebase.firestore.ServerTimestamp
import kotlinx.serialization.Serializable
import java.security.Timestamp

@Serializable
data class Lesson(
    val id: String = "",                // convenience for UI diffing
    val ownerId: String,

    val title: String,
    val description: String,
    val imageUrl: String?,

    val status: LessonStatus,
    val ratingSum: Int,
    val ratingCount: Int,

    @get:ServerTimestamp
    val createdAt: Long,
    val lastUpdated: Long,
) {
    val avgRating: Float
        get() = if (ratingCount == 0) 0f else ratingSum.toFloat() / ratingCount
}

enum class LessonStatus { Active, Archived }

File: functions/src/models/Lesson.ts
import type { FirestoreDataConverter } from "firebase-admin/firestore";
import { TS } from "./common";

export enum LessonStatus {
  Active   = "Active",
  Archived = "Archived",
}

export interface Lesson {
  ownerId: string;
  title: string;
  description: string;
  status: LessonStatus;
  ratingSum: number;
  ratingCount: number;
  createdAt: TS;
  lastUpdatedAt: TS;
}

export const lessonConverter: FirestoreDataConverter<Lesson> = {
  toFirestore: (l) => l,
  fromFirestore: (snap) => snap.data() as Lesson,
};

